src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -127,13 +140,22 @@ private static void AllocateToken(SocketAsyncContext context, out SocketAsyncEng
         {
             lock (s_lock)
             {
-                if (s_currentEngine == null)
+                engine = s_currentEngines[s_allocateFromEngine];
+                if (engine == null)
                 {
-                    s_currentEngine = new SocketAsyncEngine();
+                    s_currentEngines[s_allocateFromEngine] = engine = new SocketAsyncEngine();
                 }
 
-                engine = s_currentEngine;
-                handle = s_currentEngine.AllocateHandle(context);
+                handle = engine.AllocateHandle(context);
+
+                if (engine.IsFull)
+                {
+                    // We'll need to create a new event port for the next handle.
+                    s_currentEngines[s_allocateFromEngine] = null;
+                }
+
+                // Round-robin to the next engine
```

Comment:
I wonder if round-robin is actually the best scheme.  For a large number of sockets, I can imagine that being the case.  For a relatively small number, it seems like having one thread handle all events could end up being better.  Did you experiment at all with other schemes?